polars/polars-core/src/frame/row/transpose.rs

```diff
@@ -54,25 +61,56 @@ impl DataFrame {
                 }
                 let cols = buffers
                     .into_iter()
-                    .enumerate()
-                    .map(|(i, buf)| {
-                        let mut s = buf.into_series().cast(dtype).unwrap();
-                        s.rename(&format!("column_{i}"));
-                        s
-                    })
+                    .map(|buf| buf.into_series().cast(dtype).unwrap())
                     .collect::<Vec<_>>();
                 Ok(DataFrame::new_no_checks(cols))
             }
+        }?;
+        out.set_column_names(colnames)?;
+        match keep_names_as {
+            Some(cn) => {
+                let namecol = Utf8Chunked::new(cn, self.get_column_names());
+                out.insert_at_idx_no_name_check(0, namecol.into()).cloned()
+            }
+            None => Ok(out),
         }
     }
 
     /// Transpose a DataFrame. This is a very expensive operation.
-    pub fn transpose(&self) -> PolarsResult<DataFrame> {
+    pub fn transpose(
+        &self,
+        keep_names_as: Option<&str>,
+        column_names: Option<Either<String, Vec<String>>>,
+    ) -> PolarsResult<DataFrame> {
+        let mut df = self.clone(); // Must be owned so we get the same type if dropping a column.
+        let colnames_t = match column_names {
+            None => (0..self.height()).map(|i| format!("column_{i}")).collect(),
+            Some(cn) => match cn {
+                Either::Left(cname) => {
+                    let new_names = self.column(&cname).and_then(|x| x.utf8())?;
+                    polars_ensure!(!new_names.has_validity(), ComputeError: "Column with new names can't have null values");
+                    df = self.drop(&cname)?;
+                    new_names
+                        .into_no_null_iter()
+                        .map(|s| s.to_owned())
+                        .collect()
+                }
+                Either::Right(names) => {
+                    polars_ensure!(names.len() == self.height(), ShapeMismatch: "Length of new column names must be the same as the row count");
+                    names
+                }
+            },
+        };
+        if let Some(cn) = keep_names_as {
+            // Check that the column name we're using for the original column names is unique before
+            // wasting time transposing
+            polars_ensure!(!colnames_t.contains(&cn.to_owned()), Duplicate: "{} is already in output column names", cn)
```

Comment:
Doesn't `cn.as_ref()` or `as_str()` work here? That saves an allocation.

Reply:
I think I tried that and it didn't work because `colnames_t` contains owned strings. I'm sure there's a better way to do it. I'm no expert in Rust strings, and they're not easy to work with to begin with.

Reply:
Right, `contains` doesn't accept the borrow trait. In that case I think we better use `!iter().any(|a| a.as_ref() == b.as_ref())`.

Reply:
That doesn't work but comparing `a.as_str()` to `cn` does.
What's the importance of one seemingly small allocation here? Does it increase the chance of cache issues later?